Data.Rmd

```diff
@@ -0,0 +1,567 @@
+  favorite_letters <<- l
+  invisible(old)
+}
+```
+
+`favorite_letters` is initialized to ("a", "b", "c") when the package is built.
+The user can then inspect `favorite_letters` with `mfl()`, at which point they'll probably want to register *their* favorite letters with `set_mfl()`.
+Note that we've used the super assignment operator `<<-` in `set_mfl()` in the hope that this will reach up into the package environment and modify the internal data object `favorite_letters`.
+But a call to `set_mfl()` fails like so:[^data-2]
+
+[^data-2]: This example will execute without error if you define `favorite_letters`, `mfl()`, and `set_mfl()` in the global workspace and call `set_mfl()` in the console.
+    But this code will fail once `favorite_letters`, `mfl()`, and `set_mfl()` are defined *inside a package*.
+
+```{r, eval = FALSE}
+mfl()
+#> [1] "a" "b" "c"
+
+set_mfl(c("j", "f", "b"))
+#> Error in set_mfl() : 
+#>   cannot change value of locked binding for 'favorite_letters'
+```
+
+Because `favorite_letters` is a regular character vector, modification requires making a copy and rebinding the name `favorite_letters` to this new value.
+And that is what's disallowed: you can't change the binding for objects in the package namespace (well, at least not without trying harder than this).
+Defining `favorite_letters` this way only works if you will never need to modify it.
+
+However, if we maintain state within an internal package environment, we **can** modify objects contained in the environment (and even add completely new objects).
+Here's an alternative implementation that uses an internal environment named "the".
+
+```{r, eval = FALSE}
+the <- new.env(parent = emptyenv())
+the$favorite_letters <- letters[1:3]
+
+#' Report my favorite letters
+#' @export
+mfl2 <- function() {
+  the$favorite_letters
+}
+
+#' Change my favorite letters
+#' @export
+set_mfl2 <- function(l = letters[24:26]) {
+  old <- the$favorite_letters
+  the$favorite_letters <- l
+  invisible(old)
+}
+```
+
+Now a user *can* register their favorite letters:
+
+```{r, eval = FALSE}
+mfl2()
+#> [1] "a" "b" "c"
+
+set_mfl2(c("j", "f", "b"))
+#> [1] "j" "f" "b"
+
+mfl2()
+#> [1] "j" "f" "b"
+```
+
+Note that this new value for `the$favorite_letters` persists only for the remainder of the current R session (or until the user calls `set_mfl2()` again).
+More precisely, the altered state persists only until the next time the package is loaded (including via `load_all()`).
+At load time, the environment `the` is reset to an environment containing exactly one object, named `favorite_letters`, with value ("a", "b", "c").
+It's like the movie Groundhog Day.
+(We'll discuss more persistent package- and user-specific data in the next section.)
+
+Jim Hester introduced our group to the nifty idea of using "the" as the name of an internal package environment.
+This lets you to refer to the objects inside in a very natural way, such as `the$token`, meaning "the token".
```

Comment:
```suggestion
This lets you to refer to the objects inside in a very natural way, such as `the$token`, meaning "_the_ token".
```

In my head I read it with emphasis on _the_ to stress that it is a singleton and it might be worth showing that like this